pkg/providers/provider.go

```diff
@@ -33,7 +33,7 @@ type Provider interface {
 	GenerateMHC() ([]byte, error)
 	ChangeDiff(currentSpec, newSpec *cluster.Spec) *types.ComponentChangeDiff
 	RunPostControlPlaneUpgrade(ctx context.Context, oldClusterSpec *cluster.Spec, clusterSpec *cluster.Spec, workloadCluster *types.Cluster, managementCluster *types.Cluster) error
-	UpgradeNeeded(ctx context.Context, newSpec, currentSpec *cluster.Spec) (bool, error)
+	UpgradeNeeded(ctx context.Context, newSpec, currentSpec *cluster.Spec, cluster *types.Cluster, datacenterConfig DatacenterConfig, machineConfigs []MachineConfig) (bool, error)
```

Comment:
I don't think we need to add here `datacenterConfig DatacenterConfig, machineConfigs []MachineConfig`

The providers can retrieve them if necessary
In fact, if you look at how the upgrade workflow is getting them, they come from the provider